kpi-tracker/src/main.rs

```diff
@@ -0,0 +1,248 @@
+    let accounts = node
+        .get_account_list(block_hash)
+        .await
+        .context(format!("Could not get accounts for block: {}", block_hash))?
+        .response;
+
+    let accounts_details_map = accounts
+        .then(|res| {
+            let mut node = node.clone();
+
+            async move {
+                let account = res?;
+                let details = account_details(&mut node, block_hash, &account).await?;
+
+                Ok((account, details))
+            }
+        })
+        .try_fold(BTreeMap::new(), |mut map, (account, details)| async move {
+            map.insert(account, details);
+            Ok(map)
+        })
+        .await;
+
+    accounts_details_map
+}
+
+/// Returns a Map of AccountAddress, AccountDetails pairs not already included in `accounts_table`
+async fn new_accounts_in_block(
+    node: &mut Client,
+    block_hash: &BlockHash,
+    accounts_table: &AccountsTable,
+) -> anyhow::Result<BTreeMap<AccountAddress, AccountDetails>> {
+    let mut transactions = node
+        .get_block_transaction_events(block_hash)
+        .await
+        .context(format!(
+            "Could not get transactions for block: {}",
+            block_hash
+        ))?
+        .response;
+
+    let mut new_accounts = BTreeMap::new();
+
+    while let Some(res) = transactions.next().await {
+        let transaction = res.context("Stream stopped prematurely")?;
+
+        match transaction.details {
+            AccountCreation(act)
+                if !accounts_table
+                    .keys()
+                    .into_iter()
+                    .any(|stored_address| act.address.is_alias(stored_address)) =>
+            {
+                let address = act.address;
+                let account_details = account_details(node, block_hash, &address).await?;
+
+                new_accounts.insert(address, account_details);
+            }
+            _ => {}
+        }
+    }
+
+    Ok(new_accounts)
+}
+
+/// Insert as a single DB transaction to facilitate easy recovery, as the service can restart from
+/// current height stored in DB.
+fn update_db(
+    db: &mut DB,
+    (block_hash, block_details): (BlockHash, BlockDetails),
+    accounts: BTreeMap<AccountAddress, AccountDetails>,
+) {
+    db.blocks.insert(block_hash, block_details);
+    db.accounts.extend(accounts.into_iter());
+}
+
+async fn handle_block(
+    node: &mut v2::Client,
+    block: FinalizedBlockInfo,
+    db: &mut DB,
+) -> anyhow::Result<()> {
+    let block_info = node
+        .get_block_info(block.block_hash)
+        .await
+        .context(format!(
+            "Could not get block info for block: {}",
+            block.block_hash
+        ))?
+        .response;
+
+    let block_details = BlockDetails {
+        block_time: block_info.block_slot_time,
+        height: block_info.block_height,
+    };
+
+    let new_accounts = if block.height.height == 0 {
+        accounts_in_block(node, &block.block_hash).await?
```

Comment:
I think it would be better (for clarity) to handle this differently.

YOu would have one function "initialize everything" (which initializes data from the genesis block) and then this `handle_block` function which just processes blocks and adds on. It will be simpler to understand that way.

Reply:
I tried to split it up so genesis block is handled once and not checked in every iteration.